src/NMonad/Operations.hs

```diff
@@ -0,0 +1,81 @@
+------------------------------------------------------------------------------------------------------------------------
+-- |
+-- Module      :  NMonad.Operations
+-- Description :  NMonad operations on internal state.
+-- Copyright   : (c) d3adb5 2023
+-- License     : BSD3
+-- Maintainer  : d3adb5 <me@d3adb5.net>
+-- Stability   : experimental
+--
+-- This module provides standard operations within the 'N' monad that have an effect on internal state.
+------------------------------------------------------------------------------------------------------------------------
+
+module NMonad.Operations
+  ( indexNotification
+  , removeNotification
+  , replaceNotification
+  , updateNotification
+  , generateNotificationId
+  , makeNotification
+  , processNotification
+  ) where
+
+import Prelude hiding (lookup)
+
+import Control.Lens (over, view)
+
+import Data.Bool.HT (if')
+import Data.Map (insert, lookup, adjust, delete)
+import Data.Word (Word32)
+
+import NMonad.Lenses (notifications, notificationCount)
+import NMonad.Core hiding (notifications, notificationCount)
+
+-- | Produce a new 'Notification' from a 'DBusNotification'.
+makeNotification :: DBusNotification -> N Notification
+makeNotification (DBusNotification appName rId appIcon summ bod acts hnts tout) = do
+  replacesIfNecessary <- asks (if' . (&& rId == 0) . not . disableReplacement . configuration)
+  notificationId <- replacesIfNecessary <$> generateNotificationId <*> pure rId
+  return def
+    { applicationName = appName
+    , identifier = notificationId
+    , applicationIcon = appIcon
+    , summary = summ
+    , body = bod
+    , actions = acts
+    , hints = hnts
```

Comment:
Are these contractions a naming choice because you cannot use "body" or "summary" as parameter names?

Reply:
It's to avoid name shadowing. Despite the `DisambiguateFieldRecords` pragma allowing GHC to differentiate between `actions` in a record update (used here) and another binding under the name `actions`, there already is a top level binding by that name exported by `NMonad.Core`: the `Lens' Notification [Text]`.

Creating another binding will result in name shadowing, and warnings will be issued by the compiler. I'd rather avoid that. Shadowing the bindings and using the `RecordWildcards` pragma could simplify this function a little bit, but I'd still have to add a `{-# OPTIONS_GHC -Wno-name-shadowing #-}` to the top of the file, like I did in `test/Test/NMonad.hs`.